extensions/amp-story/1.0/amp-story-embedded-component.js

```diff
@@ -536,37 +574,152 @@ export class AmpStoryEmbeddedComponent {
     this.resources_.measureMutateElement(target,
         /** measure */
         () => {
           const targetRect = target./*OK*/getBoundingClientRect();
+          // TODO(#20832): Store DOMRect for the page in the store to avoid
+          // having to call getBoundingClientRect().
           const pageRect = this.componentPage_./*OK*/getBoundingClientRect();
 
-          const centeredTop = pageRect.height / 2 - targetRect.height / 2;
-          const centeredLeft = pageRect.width / 2 - targetRect.width / 2;
+          // Gap on the left of the element between full-screen size and
+          // current size.
+          const leftGap = (storedStyle.width - targetRect.width) / 2;
+          // Distance from left of page to what will be the left of the
+          // element in full-screen.
+          const fullScreenLeft = targetRect.left - leftGap - pageRect.left;
+          const centeredLeft = pageRect.width / 2 - storedStyle.width / 2;
+          state.translateX = centeredLeft - fullScreenLeft;
+
+          // Gap on the top of the element between full-screen size and
+          // current size.
+          const topGap = (storedStyle.height - targetRect.height) / 2;
+          // Distance from top of page to what will be the top of the element in
+          // full-screen.
+          const fullScreenTop = targetRect.top - topGap - pageRect.top;
+          const centeredTop = pageRect.height / 2 - storedStyle.height / 2;
+          state.translateY = centeredTop - fullScreenTop;
+        },
+        /** mutate */
+        () => {
+          target.classList.toggle('i-amphtml-expanded-component', true);
+
+          const styleEl = embedStyleEls[embedId];
+          embedStyleEls[embedId].textContent = styleEl.textContent.split(
+              `scale(${storedStyle.scale})`).join(`scale(1)
+              translate3d(${state.translateX}px, ${state.translateY}px, 0)`);
+        });
+  }
 
-          // Only account for offset from target to page borders. Since in
-          // desktop mode page is not at the borders of viewport.
-          const leftOffset = targetRect.left - pageRect.left;
-          const topOffset = targetRect.top - pageRect.top;
+  /**
+   * Resizes expandable element before it is expanded to full-screen, in
+   * preparation for its animation. It resizes it to its full-screen size, and
+   * scales it down to match size set by publisher, adding negative margins so
+   * that content around stays put.
+   * @param {!Element} pageEl
+   * @param {!Element} element
+   * @param {!../../../src/service/resources-impl.Resources} resources
+   */
+  static prepareForAnimation(pageEl, element, resources) {
+    let elId = null;
+
+    // When a window resize happens, we must reset the styles and prepare the
+    // animation again.
+    if (element.hasAttribute(EMBED_ID_ATTRIBUTE_NAME)) {
+      elId = element.getAttribute(EMBED_ID_ATTRIBUTE_NAME);
+      embedStyleEls[elId].textContent = '';
```

Comment:
Isn't it possible that there's no element matching `elId`, which would then make this throw?

Reply:
If theres an element with the `i-amphtml-embed-id` (which we are checking for here), then it means we'll have an element matching `elId`.

At the same time we set this attribute,  we are creating a new entry in `embedStyleEls` with the corresponding id:
https://github.com/ampproject/amphtml/pull/20683/files#diff-57a44e86ba2ed73cb5a28c4d97f17a1cR709

Reply:
Okay, maybe we can add...

```javascript
const embedStyleEl = dev().assertElement(embedStyleEls[elId],
    `Failed to look up embed style element with ID ${elId}`);
```

...and then use `embedStyleEl` for all future operations?

As it's written now, if we ever have a bug or edge case where the `embedStyleEls` object gets out of sync with the DOM representation, it will just surface as `Cannot read property 'textContent' of undefined`, and we'll have no idea what that means 😄

Reply:
Done.